rust/protocol/src/address.rs

```diff
@@ -1,26 +1,74 @@
 //
-// Copyright 2020 Signal Messenger, LLC.
+// Copyright 2020-2021 Signal Messenger, LLC.
 // SPDX-License-Identifier: AGPL-3.0-only
 //
 
+#![warn(missing_docs)]
+
+//! A normalized representation of an individual Signal client instance.
+
+#[cfg(doc)]
+use crate::SignalMessage;
+
 use std::fmt;
 
+/// The type used in memory to represent a *device*, i.e. a particular Signal client instance which
+/// represents some user.
+///
+/// Used in [ProtocolAddress].
+pub type DeviceId = u32;
+
+/// Represents a unique Signal client instance as `(<user ID>, <device ID>)` pair.
 #[derive(Clone, Debug, Hash, Eq, PartialEq, PartialOrd, Ord)]
 pub struct ProtocolAddress {
     name: String,
-    device_id: u32,
+    device_id: DeviceId,
 }
 
 impl ProtocolAddress {
-    pub fn new(name: String, device_id: u32) -> Self {
+    /// Create a new address.
+    ///
+    /// - `name` defines a user's public identity, and therefore must be globally unique to that
+    /// user.
+    /// - Each Signal client instance then has its own `device_id`, which must be unique among
+    ///   all clients for that user.
+    ///
+    ///```
+    /// use libsignal_protocol::{DeviceId, ProtocolAddress};
+    /// # use rand::{Rng, rngs::OsRng};
+    /// # use uuid::Uuid;
+    ///
+    /// // This is a unique id for some user, typically a UUID.
+    /// let user_id: String;
+    /// # user_id = Uuid::new_v4().to_hyphenated().to_string();
+    /// // This device id would be associated with a particular client instance representing
+    /// // the user.
+    /// let device_id: u32;
+    /// # device_id = (&mut OsRng).gen();
+    /// let address = ProtocolAddress::new(user_id.clone(), device_id.into());
+    ///
+    /// assert!(address.name() == &user_id);
+    /// assert!(address.device_id() == device_id.into());
+    ///```
+    pub fn new(name: String, device_id: DeviceId) -> Self {
         ProtocolAddress { name, device_id }
     }
 
+    /// A unique identifier for the target user.
+    ///
+    /// In the Signal app, this is usually a UUID, although that may change in the future.
```

Comment:
This codebase is likely to get stricter on this rather than looser, either by validating that the name is a UUID or perhaps even parsing it into the cheaper `uuid::Uuid` bytes form. It's probably better not to speculate about the Signal app at all here, and just stop at "this is usually a UUID".

Reply:
Thanks! That's easy enough to follow.